File: tests/test_tx_port.py
#!/usr/bin/env python3
"""
Test script to verify TX port functionality
This script tests the TX port connection and data transmission to /dev/ttyACM0 at 19200 baud
"""

import serial
import struct
import time
import sys

def create_test_packet():
    """Create a test telemetry packet"""
    packet = bytearray()
    
    # Start delimiter
    packet.append(0xFF)
    packet.append(0xFF)
    
    # Team ID and packet counter
    packet.append(142)  # Team ID
    packet.append(1)    # Packet counter
    
    # Test data (floats)
    test_values = [
        100.5,    # altitude_agl
        41.015234, # rocket_latitude
        28.979530, # rocket_longitude
        120.3,    # rocket_gps_altitude
        41.015240, # payload_latitude
        28.979535, # payload_longitude
        115.7,    # payload_gps_altitude
        0.5,      # gyro_x
        -0.3,     # gyro_y
        0.8,      # gyro_z
        0.1,      # accel_x
        -0.2,     # accel_y
        -9.81,    # accel_z
        15.5,     # angle
    ]
    
    # Pack all float values
    for value in test_values:
        packet.extend(struct.pack('<f', value))
    
    # Status code
    packet.append(2)  # Status code
    
    # Calculate checksum
    checksum = 0
    for i in range(2, len(packet)):  # From team_id to status_code
        checksum ^= packet[i]
    packet.append(checksum)
    
    # End delimiter
    packet.append(0x0D)
    packet.append(0x0A)
    
    return bytes(packet)

def create_simple_test_packet():
    """Create a simple test packet"""
    packet = bytearray()
    
    # Header
    packet.append(0xAB)
    
    # 8 float values (32 bytes)
    values = [
        100.5,    # altitude
        1001.25,  # pressure
        0.1,      # accel_x
        -0.2,     # accel_y
        -1.0,     # accel_z
        0.5,      # gyro_x
        -0.3,     # gyro_y
        15.5,     # angle
    ]
    
    for value in values:
        packet.extend(struct.pack('<f', round(value, 2)))
    
    # Checksum
    checksum = sum(packet[1:]) % 256
    packet.append(checksum)
    
    # Footer
    packet.append(0x0D)
    packet.append(0x0A)
    
    return bytes(packet)

def test_tx_port():
    """Test TX port connection and data transmission"""
    port = "/dev/ttyACM0"
    baud = 19200
    
    print(f"🔌 Testing TX port connection to {port} @ {baud} baud...")
    
    try:
        # Try to connect to the port
        ser = serial.Serial(
            port=port,
            baudrate=baud,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            stopbits=serial.STOPBITS_ONE,
            timeout=1
        )
        
        print(f"✅ Successfully connected to {port}")
        
        # Test 1: Send simple packet
        print("\n📤 Test 1: Sending simple packet...")
        simple_packet = create_simple_test_packet()
        ser.write(simple_packet)
        ser.flush()
        print(f"✅ Sent simple packet ({len(simple_packet)} bytes)")
        
        # Show packet contents
        packet_hex = ' '.join([f"{b:02X}" for b in simple_packet])
        print(f"📦 Packet: {packet_hex}")
        
        time.sleep(0.5)
        
        # Test 2: Send full packet
        print("\n📤 Test 2: Sending full packet...")
        full_packet = create_test_packet()
        ser.write(full_packet)
        ser.flush()
        print(f"✅ Sent full packet ({len(full_packet)} bytes)")
        
        # Show packet contents
        packet_hex = ' '.join([f"{b:02X}" for b in full_packet[:20]])
        print(f"📦 Packet: {packet_hex}...")
        
        time.sleep(0.5)
        
        # Test 3: Send multiple packets
        print("\n📤 Test 3: Sending multiple packets...")
        for i in range(5):
            # Update packet counter
            full_packet = create_test_packet()
            full_packet = bytearray(full_packet)
            full_packet[3] = i + 1  # Update packet counter
            
            ser.write(bytes(full_packet))
            ser.flush()
            print(f"✅ Sent packet {i+1}/5")
            time.sleep(0.1)
        
        print("\n✅ All tests completed successfully!")
        print(f"📊 Sent packets to {port} @ {baud} baud")
        
        ser.close()
        return True
        
    except serial.SerialException as e:
        print(f"❌ Serial connection failed: {e}")
        print("💡 Make sure:")
        print("   - The device is connected to /dev/ttyACM0")
        print("   - You have permission to access the port")
        print("   - The port is not being used by another application")
        return False
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        return False

def list_available_ports():
    """List all available serial ports"""
    import serial.tools.list_ports
    
    print("🔍 Available serial ports:")
    ports = serial.tools.list_ports.comports()
    
    if not ports:
        print("   No serial ports found")
        return
    
    for port in ports:
        print(f"   {port.device}: {port.description}")
        if port.hwid:
            print(f"      Hardware ID: {port.hwid}")

if __name__ == "__main__":
    print("🧪 TX Port Test Script")
    print("=" * 50)
    
    # List available ports first
    list_available_ports()
    print()
    
    # Test TX port
    success = test_tx_port()
    
    if success:
        print("\n🎉 TX port test PASSED!")
        sys.exit(0)
    else:
        print("\n💥 TX port test FAILED!")
        sys.exit(1) 

File: tests/test_tx_port.js
#!/usr/bin/env node
/**
 * test_tx_port.js
 * Tests TX serial port functionality by sending test packets.
 * This test sends telemetry packets through the serial port to verify
 * hardware communication with the rocket's transmission system.
 * JS port of Tests/test_tx_port.py
 */

'use strict';

const { SerialPort } = require('serialport');

const TX_PORT = '/dev/ttyACM0';
const BAUD    = 19200;

// ─── Packet builders ──────────────────────────────────────────────────────────

function createTestPacket() {
  const packet = Buffer.alloc(78, 0);
  let off = 0;

  // Header
  packet[off++] = 0xff; packet[off++] = 0xff;
  packet[off++] = 0x54; packet[off++] = 0x52;

  packet[off++] = 142; // Team ID
  packet[off++] = 1;   // Counter

  const values = [
    100.5, 41.015234, 28.979530, 120.3,
    41.015240, 28.979535, 115.7,
    0.5, -0.3, 0.8, 0.1, -0.2, -9.81, 15.5,
    0, 0, 0,
  ];
  for (const v of values) {
    packet.writeFloatLE(v, off);
    off += 4;
  }

  packet[off++] = 2; // Status

  // Checksum (XOR from offset 4 to current-1)
  let cs = 0;
  for (let i = 4; i < off; i++) cs ^= packet[i];
  packet[off++] = cs;

  packet[off++] = 0x0d;
  packet[off]   = 0x0a;

  return packet;
}

function createSimpleTestPacket() {
  const packet = Buffer.alloc(42, 0);
  let off = 0;

  packet[off++] = 0xab; // Header

  const values = [100.5, 1001.25, 0.1, -0.2, -1.0, 0.5, -0.3, 15.5];
  for (const v of values) {
    packet.writeFloatLE(Math.round(v * 100) / 100, off);
    off += 4;
  }

  let cs = 0;
  for (let i = 1; i < off; i++) cs = (cs + packet[i]) & 0xff;
  packet[off++] = cs;

  packet[off++] = 0x0d;
  packet[off]   = 0x0a;

  return packet.slice(0, off + 1);
}

// ─── Port listing ─────────────────────────────────────────────────────────────

async function listAvailablePorts() {
  const ports = await SerialPort.list();
  console.log('🔍 Available serial ports:');
  if (!ports.length) { console.log('   (none found)'); return; }
  for (const p of ports) {
    console.log(`   ${p.path}: ${p.manufacturer ?? p.friendlyName ?? '—'}`);
    if (p.pnpId) console.log(`      HW ID: ${p.pnpId}`);
  }
}

// ─── TX test ─────────────────────────────────────────────────────────────────

function testTxPort() {
  return new Promise((resolve) => {
    console.log(`\n🔌 Testing TX port ${TX_PORT} @ ${BAUD} baud...`);

    const ser = new SerialPort({ path: TX_PORT, baudRate: BAUD, autoOpen: false });

    ser.open((err) => {
      if (err) {
        console.error(`❌ Serial connection failed: ${err.message}`);
        console.log('💡 Make sure:');
        console.log(`   - Device is connected to ${TX_PORT}`);
        console.log('   - You have port access permissions');
        console.log('   - Port is not in use by another process');
        return resolve(false);
      }

      console.log(`✅ Connected to ${TX_PORT}`);

      // Test 1 – simple packet
      const simple = createSimpleTestPacket();
      ser.write(simple, () => {
        const hex = [...simple].map(b => b.toString(16).padStart(2,'0').toUpperCase()).join(' ');
        console.log(`\n📤 Test 1: Sent simple packet (${simple.length} bytes)`);
        console.log(`📦 Packet: ${hex}`);
      });

      setTimeout(() => {
        // Test 2 – full 78-byte packet
        const full = createTestPacket();
        ser.write(full, () => {
          const preview = [...full.slice(0,20)].map(b => b.toString(16).padStart(2,'0').toUpperCase()).join(' ');
          console.log(`\n📤 Test 2: Sent full packet (${full.length} bytes)`);
          console.log(`📦 Packet: ${preview}...`);
        });

        // Test 3 – burst of 5
        let sent = 0;
        const burst = () => {
          if (sent >= 5) {
            ser.close();
            console.log('\n✅ All tests completed!');
            console.log(`📊 Sent packets to ${TX_PORT} @ ${BAUD} baud\n`);
            return resolve(true);
          }
          const p = createTestPacket();
          p[5] = sent + 1; // patch counter
          ser.write(p, () => {
            console.log(`✅ Burst packet ${sent + 1}/5`);
            sent++;
            setTimeout(burst, 100);
          });
        };

        setTimeout(burst, 600);
      }, 500);
    });

    ser.on('error', (err) => { console.error(`❌ Serial error: ${err.message}`); resolve(false); });
  });
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log('🧪 TX Port Test Script (JS)');
  console.log('='.repeat(50));

  await listAvailablePorts();

  const ok = await testTxPort();
  console.log(ok ? '\n🎉 TX port test PASSED!' : '\n💥 TX port test FAILED!');
  process.exit(ok ? 0 : 1);
}

main().catch(console.error);